### PR TITLE
docs: Autogenerate version history rst from yaml changelogs

### LINKED ISSUE #20506: Single backticks are no longer being caught in version_history
The code in `tools/docs/rst_check.py` is a bit brittle and is earmarked to be moved/improved to pytooling fairly soon

One thing that seems to have stopped working is the single backtick checker, and as a result we are getting quite a few PRs landing with only single backticks

Ideally we dont do this with regexes - altho in this particular case i havent found any of the static rst(check/lint) tools to work - i think probably this could be done with a sphinx plugin at render time

Either way, i think its quite important that we get this working again ASAP as it leads to bad docs and PR noise

### LINKED ISSUE #21065: Use structured data (ie yaml) for changelog (version histories)
### description

Surfacing a conversation on slack with @mattklein123 ...

When adding a changelog entry, instead of editing `current.rst` as we do now we could instead add entries into a eg `changelog.yaml` file

The `current.rst` file could then be built from the yaml

This has the advantage that it would make it easier to validate/manipulate the data in the yaml file

The changelog entries would still want to be valid `rst` 

If we go down this path one question it throws up is how we store historical changelogs - ie do we just generate the `v1.X.Y.rst` file and store it as we do now, or do we retain the yaml files and leave them to be built/rendered in bazel

if we take the latter path we probs want to shift existing rst -> `changelog.v1.X.Y.yaml` and make the version_history rst entirely dynamic and not just `current.rst`

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:
Additional Description:

This PR does quite a few things

- shifts changelog/version history data `rst` -> `yaml`
- adds tooling to generate rst from yaml
- reorganises layout of version history rst - current/archived/stable, grouped by minor version
- many cleanups of changelog/rst - punctuation, literals, etc
- removes versioned links from changelogs - these are now added dynamically, and will _not_ need to be editted when merging changelogs from release branches
- adds (mising) tags for early changelogs 
- adds redirects for old version history paths eg `version_history/v1.2.3.rst` -> `version_history/v1.2/v1.2.3.rst`
- renames `rst_check` -> `versions_check`
- fixes currently broken (rst) checks in ^^, and adds new version checks
- removes pytesting for this checker (these will be readded when this is moved to pytooling)
- runs changelog/rst checks against _all_ changelogs, not just current
- adds a `Project` lib for managing versioning/changelogs (this can also be useful for "opening"/"closing" release branches and syncing changelogs from other release branches)

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] Fix #21065 Fix #20506
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: tools/docs/BUILD [hunk 1]
```diff
@@ -1,6 +1,6 @@
-load("@rules_python//python:defs.bzl", "py_binary")
```
Comment (phlax) [inline tools/docs/BUILD:1]: python tooling changes start here

**Comment by phlax** [comment]
this can work slightly differently in each of these situations:

- non-dev release branch (ie published docs)
- dev on main
- dev on a release branch

the latter 2 work in a similar way, but require slighly different codepaths to achieve it

im flagging this because we need to ensure this code works for the pubslishing situation, not just in dev

i have clicky tested it pretty thoroughly now - when i shift this code to pytooling i will add some integration tests

**Comment by phlax** [comment]
new checks can be seen here https://dev.azure.com/cncf/envoy/_build/results?buildId=107395&view=logs&j=c5dd2866-6ab3-5f3c-3a44-4cef0ec909b5&t=a9eb66d6-8944-5769-b3f7-476949dadcb8&l=7722

File: tools/docs/generate_version_histories.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            dev_version=self.project.dev_version,
+            changelogs=self.project.changelogs,
+            minor_versions=self.project.minor_versions,
+            stable_versions=self.project.stable_versions,
+            archived_versions=self.project.archived_versions)
+        self.tpath.joinpath("version_history.rst").write_text(f"{version_history_rst.strip()}\n\n")
+
+    def write_version_history_minor_indeces(self) -> None:
+        for i, (minor_version, patches) in enumerate(self.project.minor_versions.items()):
+            if self.project.is_main_dev and i == 0:
```
Comment (phlax) [inline tools/docs/generate_version_histories.py:237]: probs this should be any dev version

File: tools/docs/versions_check.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            if v.release_date == "Pending"
+            and k.base_version != self.project.current_version.base_version
+        ])
+        if not extra_pending:
+            self.succeed("pending", [f"No extraneous pending versions found"])
+        elif self.project.current_version.is_devrelease:
+            self.error(
+                "pending", [f"Only current version should be pending, found: {extra_pending}"])
+        else:
+            self.error("pending", [f"Nothing should be pending, found: {extra_pending}"])
```
Comment (phlax) [inline tools/docs/versions_check.py:155]: this wont be hit

File: tools/docs/generate_version_histories.py (hunk 2 above)
Comment (phlax) [reply tools/docs/generate_version_histories.py:237]: actualy this is correct - only main dev means that the minor version is not added

**Comment by mattklein123** [comment]
This is amazing. Thank you @phlax! I looked at the docs and at a high level it LGTM. Did you do any diffing to check for changes between what we have now and what the new tooling generates?

**Review by mattklein123 — APPROVED** [review]

**Comment by phlax** [comment]
> I looked at the docs and at a high level it LGTM. Did you do any diffing to check for changes between what we have now and what the new tooling generates?

i had some "do-once" type tooling, but mostly i decided to do it ~by-hand so i could clean up as i went

i was v v careful about how i migrated the data, so reasonably confident that the migration is correct

also, at the end i fixed the existing tooling to run on all files, so i did some further cleanups with that

**Comment by lizan** [comment]
@phlax You might want to also update https://github.com/envoyproxy/envoy/blob/main/.gitattributes#L1 for those yaml files.

**Review by mattklein123 — APPROVED** [review]
Nice!

**Pull request merged**